### DISCUSSION TITLE
Feeding `request.stream` into `tarfile.open(fileobj=...)`

### DISCUSSION THREAD

**Discussion by bbeattie-phxlabs** [fastapi/fastapi#5582] [Questions] [answered]
I’m looking to use Starlette’s request.stream from a streaming upload and feed it straight into a tarfile.open similar to `tarfile.open(fileobj=request.stream(), mode="r|")`.

But this doesn’t work as [request.stream() is an async generator](https://github.com/encode/starlette/blob/master/starlette/requests.py#L208) and `tarfile.open` is expecting a file-like object (e.g. io.BufferedReader). So this understandably errors out with

```
python3.10/tarfile.py", line 560, in __read
    buf = self.fileobj.read(self.bufsize)
AttributeError: 'async_generator' object has no attribute 'read'
```

Anyone have experiencing with converting an asynchronous generator to an `io.BufferedReader`? Or any other good way of supporting streaming uploads directly into `tarfile.open`?

Note that https://github.com/tiangolo/fastapi/issues/4585 exists, but that ticket is only in the context of an UploadFile, which I can't use as I need to do streaming upload of the file on the client-side.

**Comment by bradbeattie** [comment, marked as answer]
For anyone reading after-the-fact, a solution I found was to

* iterate over the `request.stream()`, putting the blocks into a queue
* have a separate thread that consumes the queue yielding bytes
* convert the generator into a buffered reader as per https://stackoverflow.com/a/51546783
* pass the buffered reader into `tarfile.open(fileobj=...)`

Not exactly _elegant_, but it gets the job done.
